README.md

```diff
@@ -0,0 +1,400 @@
+```rs
+unsafe impl<trait Trait, T: Trait> Unsize<dyn Trait> for T {
+    fn target_metadata(metadata: <Self as Pointee>::Metadata) -> <dyn Trait as Pointee>::Metadata {
+        // magic
+    }
+}
+```
+
+For the unstable trait upcasting feature, the compiler will generate the following `Unsize` implementations:
+```rs
+unsafe impl<trait Trait, trait Super> Unsize<dyn Super> for dyn Trait
+where
+    dyn Trait: Super
+{
+    unsafe fn target_metadata(metadata: <Self as Pointee>::Metadata) -> <dyn super as Pointee>::Metadata {
+        // compiler magic
+    }
+}
+```
+This is safe to do, as the metadata of source can be safely extracted from a raw pointer without touching the data for a `CoerceUnsized` implementation on `*const T`/`*mut T`.
+
+
+To keep backwards compatibility (as these are already observable in today's stable rust), the compiler also generates `Unsize<Foo<..., U, ...>>` implementations for structs `Foo<..., T, ...>` if all of these conditions are met:
+- `T: Unsize<U>`.
+- Only the last field of `Foo` has a type involving `T`.
+- `Bar<T>: Unsize<Bar<U>>`, where `Bar<T>` stands for the actual type of that last field.
+- `Foo<U>: Pointee<Metadata = <Bar<U> as Pointee>::Metadata>`
+
+## Builtin unsizes for performance
+
+The compiler may "un-lower" some known unsize coercions back into builtin operations in the MIR as to not degrade performance too much, as lowering this new definition will introduce a lot of new operations that don't exist in the current unsizing logic.
+This would be similar to how builtin operators for primitives work currently, where they are typechecked with the trait impls but then lowered back to builtin operators in the mir.
+
+## TypeMetadata<T> and Unsizing
+
+See the following PR for context: https://github.com/rust-lang/rust/pull/97052.
+
+With this new definition, we can implement `CoerceUnsized` for `TypeMetadata` without having to special case it in the compiler as follows:
+
+```rs
+struct TypedMetadata<T: ?Sized>(pub <T as core::ptr::Pointee>::Metadata);
+
+
+impl<T, U> CoerceUnsized<TypedMetadata<U>> for TypedMetadata<T>
+where
+    T: ?Sized + Unsize<U>,
+    U: ?Sized,
+{
+    fn coerce_unsized(self) -> TypedMetadata<U> {
+        TypedMetadata(Unsize::target_metadata(self.0))
+    }
+}
+```
+
+## Pin Unsoundness
+
+See the following issue for context: https://github.com/rust-lang/rust/issues/68015
+
+The design of the new traits here do not address the underlying issue in regards to `Pin`.
+The author of this RFC feels like addressing the `Pin` soundness in the definitions of the traits is wrong, as in almost all cases where are a user implements one of these traits `Pin` will be irrelevant to them. And at its core, unsizing is not coupled with `Pin` whatsoever.
+It would make much more sense to fix the unsound `CoerceUnsized` implementation that `Pin` provides.
+That is given the current implementation of (with the new definition of the trait):
```

Comment:
```suggestion
The author of this RFC feels like addressing the `Pin` soundness in the definitions of the traits is wrong, as in almost all cases where a user implements one of these traits `Pin` will be irrelevant to them. And at its core, unsizing is not coupled with `Pin` whatsoever.

It would make much more sense to fix the unsound `CoerceUnsized` implementation that `Pin` provides.

This is the current implementation (using the new trait from this RFC):
```